### PR TITLE
Add a `spacy benchmark speed` subcommand

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This subcommand reports the mean batch performance of a model on a data set with a 95% confidence interval. For reliability, it first performs some warmup rounds. Then it will measure performance on batches with randomly shuffled documents.

To avoid having too many spaCy commands, `speed` is a subcommand of `benchmark` and `benchmark accuracy` is now an alias of `evaluate`.

Out of draft:

- [x] We need to decide how to handle that `accuracy` is now a subcommand.
- [x] I need to check/update the documentation and document the new subcommand.
- [x] ~Should probably target spaCy 4, since the module `evaluate` was renamed to `evaluate_accuracy`.~

### Types of change

New feature.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
Is there no way to keep the current `spacy evaluate` command?

**Comment by danieldk** [comment]
> Hmm, `spacy evaluate` already includes a minimal `SPEED`. Could the extended version somehow be a flag instead? I realize it's tricky with the additional options.

It really works differently with the eval loop and stuff, it seems a bit bolted-on?

> But I dislike breaking all existing uses of `spacy evaluate` if we don't have to. I might propose:
> 
> * `spacy evaluate` adds a flag for this that uses the defaults
> * a separate command lets you run it with the extended options

So rather than adding a flag, what about:

* Keep `evaluate` as-is.
* Put the speed test under `benchmark speed`.
* Make `benchmark accuracy` an alias for `evaluate`.

That way we can move to subcommands without breaking the current interface? Not sure if `benchmark` is the best name, but you get the drift (maybe `analyze`?).

**Comment by shadeMe** [comment]
> That way we can move to subcommands without breaking the current interface? Not sure if benchmark is the best name (maybe analyze?).

IMO, the `speed` subcommand for `benchmark` seems fairly superfluous (in this context, at the least). Can't we simply stick to an additional `benchmark` command and call it a day? That way, the default associations of the words `evaluate` -> `accuracy` and `benchmark` -> `speed` will still hold.

**Comment by adrianeboyd** [comment]
`evaluate` already includes the simple `SPEED`, and "benchmark" doesn't have an "only speed" association for me?

**Comment by danieldk** [comment]
> IMO, the speed subcommand for benchmark seems fairly superfluous (in this context, at the least). Can't we simply stick to an additional benchmark command and call it a day?

We discussed this a bit the other day, and one of the ideas was to have a suite of evaluation-related subcommands like we have `debug` now. `benchmark {accuracy,speed}` could be the first subcommands. The alias would be to preserve backwards compatibility, so that `spacy evaluate` still works. I think the downside of adding a top-level command (regardless of the name) is that we clutter the interface too much.

**Comment by danieldk** [comment]
> Daniël's suggestion sounds okay to me. I'm not sure that it's easy to have a 100% alias, but I'm sure there's some way to get close.

Changed in this way, also made `benchmark accuracy` an alias for `evaluate`. I think this can now target `master`, since there are no changes to existing interfaces anymore. I still need to update the docs.

**Comment by danieldk** [comment]
Added docs and moved this out of draft.

File: website/docs/api/cli.md [hunk 1]
```diff
@@ -1162,6 +1163,38 @@ $ python -m spacy evaluate [model] [data_path] [--output] [--code] [--gold-prepr
...
+| `model`              | Pipeline to benchmark the speed of. Can be a package or a path to a data directory. ~~str (positional)~~ |
+| `data_path`          | Location of benchmark data in spaCy's [binary format](/api/data-formats#training). ~~Path (positional)~~ |
+| `--batch-size`, `-b` | Set the batch size. If not set, the pipeline's batch size is used. ~~Optional[int] \(option)~~           |
+| `--no-shuffle`       | Do not shuffle documents in the benchmark data. ~~bool (flag)~~                                          |
+| `--gpu-id`, `-g`     | GPU to use, if any. Defaults to `-1` for CPU. ~~int (option)~~                                           |
+| `--batches`          | Number of batches to benchmark on. Defaults to `50`. ~~Optional[int] \(option)~~                         |
+| `--warmup`, `-w`     | Iterations over the benchmark data for warmup. Defaults to `3` ~~Optional[int] \(option)~~               |
+| `--help`, `-h`       | Show help message and available arguments. ~~bool (flag)~~                                               |
+| **CREATES**          | Pipeline speed with 95% confidence interval.                                                             |
```
Comment (shadeMe) [inline website/docs/api/cli.md:1197]: `**REPORTS** ` might be a better verb here, no?

Comment (danieldk) [reply website/docs/api/cli.md]: Good catch. Changed to `**PRINTS**`, since that verb is used for some other CLI commands.

**Review by svlandeg — COMMENTED** [review]
We should add some tests that ensure `spacy evaluate` and `spacy benchmark accuracy` both work as expected.

Cf https://github.com/explosion/spaCy/blob/master/spacy/tests/test_cli_app.py -> We can use `CliRunner` for this :-)

File: spacy/cli/evaluate.py [hunk 2]
```diff
@@ -7,12 +7,14 @@
...
-from ._util import app, Arg, Opt, setup_gpu, import_code
+from ._util import app, Arg, Opt, setup_gpu, import_code, benchmark_cli
 from ..scorer import Scorer
 from .. import util
 from .. import displacy
 
-
+@benchmark_cli.command(
+    "accuracy",
+)
```
Comment (svlandeg) [inline spacy/cli/evaluate.py:17]: Oh, that's nice and clean :-)

File: website/docs/api/cli.md [hunk 3]
```diff
@@ -1163,6 +1164,38 @@ $ python -m spacy evaluate [model] [data_path] [--output] [--code] [--gold-prepr
...
 | **CREATES**                               | Training results and optional metrics and visualizations.                                                                                                                            |
 
+## benchmark {#benchmark new="3.5"}
+
+The `spacy benchmark` CLI includes commands for benchmarking the accuracy and
+speed of your spaCy pipelines.
+
+### accuracy {#benchmark-accuracy new="3.5" tag="command"}
+
+This subcommand is an alias for [`spacy evaluate`](#evaluate).
```
Comment (svlandeg) [inline website/docs/api/cli.md:1174]: Maybe we should also have a note over at the `evaluate` documentation? Realistically we won't be deprecating `evaluate` anytime soon, but it's always good to warn long up-front.

File: website/docs/api/cli.md [hunk 4]
```diff
@@ -1163,6 +1164,38 @@ $ python -m spacy evaluate [model] [data_path] [--output] [--code] [--gold-prepr
...
+| `model`              | Pipeline to benchmark the speed of. Can be a package or a path to a data directory. ~~str (positional)~~ |
+| `data_path`          | Location of benchmark data in spaCy's [binary format](/api/data-formats#training). ~~Path (positional)~~ |
+| `--batch-size`, `-b` | Set the batch size. If not set, the pipeline's batch size is used. ~~Optional[int] \(option)~~           |
+| `--no-shuffle`       | Do not shuffle documents in the benchmark data. ~~bool (flag)~~                                          |
+| `--gpu-id`, `-g`     | GPU to use, if any. Defaults to `-1` for CPU. ~~int (option)~~                                           |
+| `--batches`          | Number of batches to benchmark on. Defaults to `50`. ~~Optional[int] \(option)~~                         |
+| `--warmup`, `-w`     | Iterations over the benchmark data for warmup. Defaults to `3` ~~Optional[int] \(option)~~               |
+| `--help`, `-h`       | Show help message and available arguments. ~~bool (flag)~~                                               |
+| **PRINTS**           | Pipeline speed with 95% confidence interval.                                                             |
```
Comment (svlandeg) [inline website/docs/api/cli.md:1197]:
```suggestion
| **PRINTS**           | Pipeline speed in words per second (WPS) with 95% confidence interval.                                                             |
```
(will require `prettier`, too)

File: spacy/cli/benchmark_speed.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+def print_mean_with_ci(sample: numpy.ndarray):
+    mean = numpy.mean(sample)
+    bootstrap_means = bootstrap(sample)
+    bootstrap_means.sort()
+
+    # 95% confidence interval
+    low = bootstrap_means[int(len(bootstrap_means) * 0.025)]
+    high = bootstrap_means[int(len(bootstrap_means) * 0.975)]
+
+    print(f"Mean: {mean:.1f} WPS (95% CI: {low-mean:.1f} +{high-mean:.1f})")
```
Comment (svlandeg) [inline spacy/cli/benchmark_speed.py:156]: I wouldn't assume all users know what "WPS" stands for

**Review by svlandeg — COMMENTED** [review]
I really like this feature, and I think having `evaluate` as an alias for `benchmark accuracy` makes sense and allows us to extend `benchmark` more in the future.

Had a few more comments/requests though!

File: spacy/cli/benchmark_speed.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+from ..training import Corpus
+from ._util import Arg, Opt, benchmark_cli, setup_gpu
+
+
+@benchmark_cli.command(
+    "speed",
+    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
+)
+def benchmark_speed_cli(
+    ctx: typer.Context,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/cli/benchmark_speed.py:24]: Fixed.

File: spacy/cli/benchmark_speed.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+def benchmark(
+    nlp: Language,
+    docs: List[Doc],
+    n_batches: int,
+    batch_size: int,
+    shuffle: bool,
+) -> numpy.ndarray:
+    if shuffle:
+        bench_docs = [
+            nlp.make_doc(random.choice(docs).text)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/cli/benchmark_speed.py:115]: An error is now emitted if the corpus is empty.

File: website/docs/api/cli.md (hunk 3 above)
Comment (danieldk) [reply website/docs/api/cli.md]: It's a bit weird anyway to have all the documentation under `evaluate` if `benchmark speed` is supposed to be the replacement. So I moved the docs to `benchmark speed` and changed the docs for `evaluate` to state that it is superseded and that it is an alias for compatibility.

File: website/docs/api/cli.md (hunk 4 above)
Comment (danieldk) [reply website/docs/api/cli.md]: Fixed, but I removed WPS completely (see below).

File: spacy/cli/benchmark_speed.py (hunk 5 above)
Comment (danieldk) [reply spacy/cli/benchmark_speed.py]: Changed to `words/s`.

**Comment by danieldk** [comment]
> We should add some tests that ensure `spacy evaluate` and `spacy benchmark` both work as expected.
> 
> Cf https://github.com/explosion/spaCy/blob/master/spacy/tests/test_cli_app.py -> We can use `CliRunner` for this :-)

Added a test that compares the output of `evaluate` and `benchmark accuracy` to check that the alias works correctly.

**Pull request merged**